src/scene/materials/shader-graph-node.js

```diff
@@ -0,0 +1,615 @@
+
+var id = 0;
+
+/**
+ * @private
+ * @class
+ * @name ShaderGraphNode
+ * @classdesc A Shader Graph Node class used by shader graphs
+ * @param {string} funcGlsl - shader function used by this material
+ * @param {string} declGlsl - shader declarations used by the shader function
+ */
+var ShaderGraphNode = function (funcGlsl, declGlsl) {
+    this.name = "Untitled";
+    this.id = id++;
+
+    // storage for asset references
+    this._ioPortAssetReferences = [];
+    this._glslAssetReferences = {};
+    this._subGraphAssetReferences = [];
+
+    this.graphData = {};
+
+    this.graphData.ioPorts = []; // input, output or constant variables
+
+    if (funcGlsl) {
+        this.graphData.customFuncGlsl = funcGlsl;
+        this.graphData.customDeclGlsl = declGlsl;
+        this.genCustomFuncVars();
+    } else {
+        this.graphData.subGraphs = [];
+        this.graphData.connections = [];
+    }
+};
+
+ShaderGraphNode.prototype.constructor = ShaderGraphNode;
+
+Object.assign(ShaderGraphNode.prototype, {
+
+    clone: function () {
+        var clone = new ShaderGraphNode();
+
+        clone.graphData = {};
+
+        clone.graphData.ioPorts = this.graphData.ioPorts.slice(0);
+        clone.graphData.subGraphs = this.graphData.subGraphs.slice(0);
+        clone.graphData.connections = this.graphData.connections.slice(0);
+
+        clone.graphData.customDeclGlsl = this.graphData.customDeclGlsl;
+        clone.graphData.customFuncGlsl = this.graphData.customFuncGlsl;
+
+        return clone;
+    },
+
+    updateShaderGraphUniforms: function (mat) {
+        for (var n = 0; n < this.graphData.ioPorts.length; n++) {
+            var ioPort = this.graphData.ioPorts[n];
+
+            if (ioPort.name.startsWith('IN_') || (ioPort.name.startsWith('CONST_') && ioPort.type === 'sampler2D')) {
+                var chunkId = '_' + this.id;
+
+                switch (ioPort.type) {
+                    case 'sampler2D':
+                        mat.setParameter(ioPort.name + chunkId, ioPort.valueTex);
+                        break;
+                    case 'float':
+                        mat.setParameter(ioPort.name + chunkId, ioPort.valueX);
+                        break;
+                    case 'vec2':
+                        mat.setParameter(ioPort.name + chunkId, [ioPort.valueX, ioPort.valueY]);
+                        break;
+                    case 'vec3':
+                        mat.setParameter(ioPort.name + chunkId, [ioPort.valueX, ioPort.valueY, ioPort.valueZ]);
+                        break;
+                    case 'vec4':
+                        mat.setParameter(ioPort.name + chunkId, [ioPort.valueX, ioPort.valueY, ioPort.valueZ, ioPort.valueW]);
+                        break;
+                    case 'samplerCube':
+                    default:
+                        // error
+                        break;
+                }
+            }
+        }
+    },
+
+    hasValidGraphData: function (graphRootTime) {

```

Comment:
this does not have comment about the purpose and it's currently not called, so I'm not entirely sure what its purpose is - will this called when asset with graph is loaded?

Reply:
yes - this was used by the asset handler - which will dynamically swap in asset resources - probably needed for next stage of MVP